### ISSUE TITLE
Missing type annotation for argument in decorator causes bug

### ISSUE THREAD

**Issue by nymann** [microsoft/pyright#2209]
**Describe the bug**

##### example.py

```py
import functools

def decorator(func):
    @functools.wraps(func)
    def wrapped(a):
        b = int(a)
        return func(b=b)

    return wrapped

@decorator
async def decorated_method(b: int):
    return 2 * b

async def example(a: str):
    return await decorated_method(a=a)
```
Results in the following diagnostics: `No parameter named "a"`, `Argument missing for parameter "b"`. Note that `a` is not type annotated in the decorator.

If we type annotate it, the diagnostics goes away:

##### example_pass.py

```py
import functools

def decorator(func):
    @functools.wraps(func)
    def wrapped(a: str):
        b = int(a)
        return func(b=b)

    return wrapped

@decorator
async def decorated_method(b: int):
    return 2 * b

async def example(a: str):
    return await decorated_method(a=a)
```

**To Reproduce**
Run `pyright example.py` and see the error output. Then run `pyright example_pass.py` and verify that no errors are shwon.

**Expected behavior**
Even though if the variable `a` is not type annotated, it should still be valid.

**Screenshots or Code**

Without the type annotation:
![image](https://user-images.githubusercontent.com/7005867/130120214-378d6c31-9753-43e7-b7dd-c7cfc36e8966.png)

With type annotation:
![image](https://user-images.githubusercontent.com/7005867/130120308-d0883fa7-e39b-4cb2-91c1-5ccea6d43729.png)

**VS Code extension or command-line**
Command line
```
$ pyright --version 
pyright 1.1.162
```

**Comment by nymann** [comment]
Output from the command line:

```
 $ pyright example.py
No configuration file found.
No pyproject.toml file found.
stubPath /home/knj/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/knj/example.py
  /home/knj/example.py:16:35 - error: No parameter named "a" (reportGeneralTypeIssues)
  /home/knj/example.py:16:18 - error: Argument missing for parameter "b" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 infos
Completed in 0.716sec
```

**Comment by nymann** [comment]
The problem also extends to `*args, **kwargs`.

### key word arguments
```py
import functools

def decorator(func):
    @functools.wraps(func)
    def wrapped(*_, **kwargs):
        a: str = kwargs.get("a", "1")
        b = int(a)
        return func(b=b)

    return wrapped

@decorator
def decorated_method(b: int):
    return 2 * b

def example(a: str):
    return decorated_method(a=a)
```

##### Result of pyright with the code above
```
17:29 - error: No parameter named "a" (reportGeneralTypeIssues)
17:12 - error: Argument missing for parameter "b" (reportGeneralTypeIssues)
```
### args
```py
import functools

def decorator(func):
    @functools.wraps(func)
    def wrapped(*args):
        a: str = args[0]
        b = int(a)
        return func(b=b)

    return wrapped

@decorator
def decorated_method(b: int):
    return 2 * b

def example(a: str):
    return decorated_method(a)
```

##### Result of pyright with the code above
```
17:29 - error: Argument of type "str" cannot be assigned to parameter "b" of type "int" in function "decorated_method"
```

**Comment by erictraut** [comment]
Unannotated decorator functions present a challenge for a type checker. We've implemented (and over time, refined) a number of heuristics to deal with this situation, but they are not foolproof.

The heuristic explains the behavior you are seeing in your samples. When a decorator is unannotated, pyright attempts to infer its type. If the resulting inferred type is a function whose input and return types are all unknown, we assume that the inference isn't sound. In this case, it assumes that the decorator preserves the signature of the decorated class.

In your example, the `decorator` function is completely unannotated. So is the `wrapped` function that the decorator is returning. So there's no way for a type checker in this case to infer the types of the input parameters or return type of the decorated type. If you add an annotation (e.g. indicate that the input parameter `a` in the `wrapped` function is type `str`), then it provides enough type information for pyright to apply the decorator. In your second example, if you annotate the `_` and `kwargs` parameters to be type `Any`, the type error will go away.

So this is a case where type evaluation needs a little bit of help in the form of some type annotations.

Note that mypy (another popular Python type checker) will not even attempt to infer the return type of a decorator that is not annotated. It always assumes that an unannotated decorator has no effect on the decorated function's signature. So pyright is already going way beyond what other type checkers do in this case.
